test/object-store/sync/flx_role_change.cpp

```diff
@@ -0,0 +1,861 @@
+    CHECK(results.size() == emps);
+    table_query = Query(table).equal(role_col, "manager");
+    results = Results(check_realm, table_query);
+    CHECK(results.size() == mgrs);
+    table_query = Query(table).equal(role_col, "director");
+    results = Results(check_realm, table_query);
+    CHECK(results.size() == dirs);
+}
+
+} // namespace
+
+TEST_CASE("flx: role change bootstraps", "[sync][flx][baas][role change][bootstrap]") {
+    auto logger = util::Logger::get_default_logger();
+    static std::unique_ptr<FLXSyncTestHarness> harness;
+
+    auto pause_download_builder = [](std::weak_ptr<SyncSession> weak_session, bool pause) {
+        if (auto session = weak_session.lock()) {
+            nlohmann::json test_command = {{"command", pause ? "PAUSE_DOWNLOAD_BUILDER" : "RESUME_DOWNLOAD_BUILDER"}};
+            SyncSession::OnlyForTesting::send_test_command(*session, test_command.dump())
+                .get_async([](StatusWith<std::string> result) {
+                    REQUIRE(result.is_ok());             // Future completed successfully
+                    REQUIRE(result.get_value() == "{}"); // Command completed successfully
+                });
+        }
+    };
+
+    enum BootstrapMode { NoReconnect, None, SingleMessage, SingleMessageMulti, MultiMessage, Any };
+    struct ExpectedResults {
+        BootstrapMode bootstrap;
+        size_t emps;
+        size_t mgrs;
+        size_t dirs;
+    };
+
+    enum TestState {
+        not_ready,
+        start,
+        reconnect_received,
+        session_resumed,
+        ident_message,
+        downloading,
+        downloaded,
+        complete
+    };
+
+    TestingStateMachine<TestState> state_machina(TestState::not_ready);
+    int64_t query_version = 0;
+    BootstrapMode bootstrap_mode = BootstrapMode::None;
+    size_t download_msg_count = 0;
+    size_t bootstrap_msg_count = 0;
+    bool role_change_bootstrap = false;
+    bool send_test_command = false;
+
+    auto setup_config_callbacks = [&](SyncTestFile& config) {
+        // Use the sync client event hook to check for the error received and for tracking
+        // download messages and bootstraps
+        config.sync_config->on_sync_client_event_hook = [&](std::weak_ptr<SyncSession> weak_session,
+                                                            const SyncClientHookData& data) {
+            state_machina.transition_with([&](TestState cur_state) -> std::optional<TestState> {
+                if (cur_state == TestState::not_ready || cur_state == TestState::complete)
+                    return std::nullopt;
+
+                using BatchState = sync::DownloadBatchState;
+                using Event = SyncClientHookEvent;
+                switch (data.event) {
+                    case Event::ErrorMessageReceived:
+                        REQUIRE(cur_state == TestState::start);
+                        REQUIRE(data.error_info);
+                        REQUIRE(data.error_info->raw_error_code == 200);
```

Comment:
should we compare the error_code rather than the raw_error_code?

Reply:
This is a `ProtocolErrorInfo` struct, which doesn't have `error_code`; only `raw_error_code` is available.